turnstile-example/turnstile/examples/cs478/2021-02-18/stlc.rkt

```diff
@@ -82,15 +92,56 @@
 -------------------
  [⊢ (λ- (x-) e-) ⇒  (→ T1 T2)])
 
-(define-typerule (#%datum . x) ≫
+(define-typerule #%datum
+  [(_ . n:integer) ≫
+   ------------
+   [⊢ (#%datum- . n) ⇒ Int]]
+  [(_ . b:boolean) ≫
+   ------------
+   [⊢ (#%datum- . b) ⇒ Bool]]
+  [(_ . (~literal unit)) ≫
+   ------------
+   [⊢ (#%datum- . '()) ⇒ Unit]]
+  [(_ . x) ≫
+   ------------
+   [#:error (type-error #:src #'x #:msg "Unsupported literal: ~v" #'x)]])
+
+(define-typerule (#%app e1 e2) ≫
+  [⊢ e1 ≫ e1- ⇒ (~→ T1 T2)]
+  [⊢ e2 ≫ e2- ⇐ T1]
   ---------
-  [⊢ 'x ⇒ Int])
-
-(define-typerule (+ e1 e2) ≫
-  [⊢ e1 ≫ e1- ⇐ Int]
-  [⊢ e2 ≫ e2- ⇐ Int]
-  ----------------
-  [⊢ (+- e1- e2-) ⇒ Int])
+  [⊢ (#%app- e1- e2-) ⇒ T2])
+
+(define-typerule (if cond thn els) ⇐ τ_expected ≫
+  [⊢ cond ≫ cond- ⇐ Bool]
+  [⊢ thn ≫ thn- ⇐ τ_expected]
+  [⊢ els ≫ els- ⇐ τ_expected]
+  ---------------------------
+  [⊢ (if- cond- thn- els-)])
+
+(define-typerule (succ t) ≫
+  [⊢ t ≫ t- ⇐ Int]
+  -----------------
+  [⊢ (#%app- add1- t-) ⇒ Int])
+
+(define-typerule (pred t) ≫
+  [⊢ t ≫ t- ⇐ Int]
+  -----------------
+  [⊢ (#%app- sub1- t-) ⇒ Int])
+
+(define-typerule (iszero t) ≫
+  [⊢ t ≫ t- ⇐ Int]
+  -----------------
+  [⊢ (#%app- zero?- t-) ⇒ Bool])
+
+;; NOTE Chapter 11 ;;
+
+;; FIXME how do you write derived forms?
+(define-typerule (begin2 e1 e2) ≫
+  [⊢ e1 ≫ e1- ⇒ T1]
```

Comment:
But you're better off just using `Unit` directly here, i.e. `[⊢ e1 ≫ e1- ⇐ Unit]`